File: src/collections_and_containers/cpp/stack/test/test_stack.cpp
#include "stack.h"
#include "gtest/gtest.h"

TEST(StackTest, EmptyStack) {
  Stack<int> stack;
  ASSERT_TRUE(stack.isEmpty());
}

TEST(StackTest, AddingMultipleElements) {
  Stack<int> stack;
  stack.push(1);
  stack.push(2);
  stack.push(3);
  EXPECT_EQ(stack.size(), 3);

  EXPECT_EQ(stack.pop(), 3);
  EXPECT_EQ(stack.pop(), 2);
  EXPECT_EQ(stack.pop(), 1);

  ASSERT_TRUE(stack.isEmpty());

  // adding the same elements in different order
  stack.push(3);
  stack.push(1);
  stack.push(2);

  EXPECT_EQ(stack.pop(), 2);
  EXPECT_EQ(stack.pop(), 1);
  EXPECT_EQ(stack.pop(), 3);
}

TEST(StackTest, Top) {
  Stack<int> stack;
  stack.push(1);
  stack.push(2);
  stack.push(3);

  EXPECT_EQ(stack.top(), 3);
}

TEST(StackTest, TopWhenEmpty) {
  Stack<int> stack;
  ASSERT_THROW(stack.top(), std::out_of_range);
}

TEST(StackTest, PopWhenEmpty) {
  Stack<int> stack;
  ASSERT_THROW(stack.pop(), std::out_of_range);
}

TEST(StackTest, CopyConstructor) {
  Stack<int> stack;
  stack.push(1);
  stack.push(2);
  stack.push(3);

  Stack<int> stack2(stack);
  EXPECT_EQ(stack2.pop(), 3);
  EXPECT_EQ(stack2.pop(), 2);
  EXPECT_EQ(stack2.pop(), 1);
}

TEST(StackTest, AssignmentOperator) {
  Stack<int> stack;
  stack.push(1);
  stack.push(2);
  stack.push(3);

  Stack<int> stack2;
  stack2 = stack;
  EXPECT_EQ(stack2.pop(), 3);
  EXPECT_EQ(stack2.pop(), 2);
  EXPECT_EQ(stack2.pop(), 1);
}

File: src/collections_and_containers/python/stack/test/test_stack.py
from src.stack import Stack
from unittest import TestCase
import unittest


class TestStack(TestCase):
    def test_empty_stack(self):
        stack = Stack()
        self.assertTrue(stack.empty())

    def test_adding_multiple_elements(self):
        stack = Stack()
        stack.push(1)
        stack.push(2)
        stack.push(3)
        self.assertEqual(stack.size(), 3)

        self.assertEqual(stack.pop(), 3)
        self.assertEqual(stack.pop(), 2)
        self.assertEqual(stack.pop(), 1)

        self.assertTrue(stack.empty())

        # adding the same elements in different order
        stack.push(3)
        stack.push(1)
        stack.push(2)

        self.assertEqual(stack.pop(), 2)
        self.assertEqual(stack.pop(), 1)
        self.assertEqual(stack.pop(), 3)

    def test_top(self):
        stack = Stack()
        stack.push(1)
        stack.push(2)
        stack.push(3)

        self.assertEqual(stack.top(), 3)

    def test_top_when_empty(self):
        stack = Stack()
        self.assertRaises(IndexError, stack.top)

    def test_pop_when_empty(self):
        stack = Stack()
        self.assertRaises(IndexError, stack.pop)

    def test_push_when_full(self):
        stack = Stack(2)
        stack.push(1)
        stack.push(2)
        self.assertRaises(IndexError, stack.push, 3)


if __name__ == "__main__":
    unittest.main()
